demo/popcha-client/web/src/authentication.py

```diff
@@ -0,0 +1,118 @@
+"""
+This file includes all the authentication logic. No external server logic is
+done here.
+"""
+
+# allows class names to be used as return types inside the class
+# see e.g. https://stackoverflow.com/a/61544901
+from __future__ import annotations
+
+import secrets
+import time
+from urllib import parse
+
+import jwt
+from jwt import PyJWTError
+from werkzeug.datastructures import MultiDict
+
+
+class Authentication:
+    """
+    Class that allows users to authenticate
+    """
+
+    def __init__(self, providers: list[dict[str, str]] = None):
+        if providers is None:
+            providers = []
+
+        self.providers: list[dict[str, str]] = providers
+        self.login_states: dict[str, (str, str, float)] = {}
```

Comment:
I think it might be nice to add a tiny bit of documentation to know why the format of these parameters is done this way.